File: tt_metal/impl/allocator/allocation_client.cpp
// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
//
// SPDX-License-Identifier: Apache-2.0

#include "allocation_client.hpp"
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#include <cstring>
#include <cstdlib>
#include <chrono>
#include <iostream>
#include <cerrno>

#define TT_ALLOC_SERVER_SOCKET "/tmp/tt_allocation_server.sock"

namespace tt::tt_metal {

// Message protocol matching allocation_server_poc.cpp
struct __attribute__((packed)) AllocMessage {
    enum Type : uint8_t {
        ALLOC = 1,
        FREE = 2,
        QUERY = 3,
        RESPONSE = 4,
        DUMP_REMAINING = 5,
        DEVICE_INFO_QUERY = 6,
        DEVICE_INFO_RESPONSE = 7,
        CB_ALLOC = 8,       // Circular buffer allocation
        CB_FREE = 9,        // Circular buffer free
        KERNEL_LOAD = 10,   // Kernel loaded
        KERNEL_UNLOAD = 11  // Kernel unloaded
    };

    Type type;            // 1 byte
    uint8_t pad1[3];      // 3 bytes padding
    int32_t device_id;    // 4 bytes
    uint64_t size;        // 8 bytes
    uint8_t buffer_type;  // 1 byte
    uint8_t pad2[3];      // 3 bytes padding
    int32_t process_id;   // 4 bytes
    uint64_t buffer_id;   // 8 bytes
    uint64_t timestamp;   // 8 bytes

    // Response fields (unused by client for alloc/free)
    uint64_t dram_allocated;
    uint64_t l1_allocated;
    uint64_t l1_small_allocated;
    uint64_t trace_allocated;
    uint64_t cb_allocated;      // NEW: Circular buffer memory
    uint64_t kernel_allocated;  // NEW: Kernel code memory

    // Device info fields (unused by client for alloc/free, used by server for device queries)
    uint64_t total_dram_size;
    uint64_t total_l1_size;
    uint32_t arch_type;  // 0=Invalid, 1=Grayskull, 2=Wormhole_B0, 3=Blackhole, 4=Quasar
    uint32_t num_dram_channels;
    uint32_t dram_size_per_channel;
    uint32_t l1_size_per_core;
    uint32_t is_available;
    uint32_t num_devices;

    // Ring buffer stats (for real-time kernel L1 usage)
    uint32_t ringbuffer_total_size;    // Total ring buffer capacity (~67KB)
    uint32_t ringbuffer_used_bytes;    // Currently used by cached kernels
    uint32_t ringbuffer_num_programs;  // Number of programs cached
    uint32_t pad3;                     // Padding for alignment
    // Total: 1+3+4+8+1+3+4+8+8+48+16+16 = 128 bytes
};

AllocationClient::AllocationClient() : socket_fd_(-1), enabled_(false), connected_(false) {
    // Check environment variable
    const char* env_enabled = std::getenv("TT_ALLOC_TRACKING_ENABLED");
    if (env_enabled && std::string(env_enabled) == "1") {
        enabled_ = true;

        // Attempt to connect (lazy, will retry on first use)
        connect_to_server();
    }
}

AllocationClient::~AllocationClient() {
    if (socket_fd_ >= 0) {
        close(socket_fd_);
        socket_fd_ = -1;
    }
}

AllocationClient& AllocationClient::instance() {
    static AllocationClient inst;
    return inst;
}

bool AllocationClient::connect_to_server() {
    if (connected_) {
        return true;
    }

    socket_fd_ = socket(AF_UNIX, SOCK_STREAM, 0);
    if (socket_fd_ < 0) {
        return false;
    }

    struct sockaddr_un addr;
    memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, TT_ALLOC_SERVER_SOCKET, sizeof(addr.sun_path) - 1);

    if (connect(socket_fd_, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        close(socket_fd_);
        socket_fd_ = -1;

        // Only warn once
        static bool warned = false;
        if (!warned) {
            std::cerr << "[TT-Metal] Warning: Allocation tracking enabled but server not available at "
                      << TT_ALLOC_SERVER_SOCKET << std::endl;
            warned = true;
        }
        return false;
    }

    // Increase socket buffer sizes to handle bursts from multiple devices
    // This significantly reduces blocking frequency during rapid allocations
    int sndbuf_size = 1024 * 1024;  // 1MB send buffer
    if (setsockopt(socket_fd_, SOL_SOCKET, SO_SNDBUF, &sndbuf_size, sizeof(sndbuf_size)) < 0) {
        // Non-fatal, just log
        std::cerr << "[TT-Metal] Warning: Could not set socket send buffer size" << std::endl;
    }

    connected_ = true;
    return true;
}

void AllocationClient::send_allocation_message(int device_id, uint64_t size, uint8_t buffer_type, uint64_t buffer_id) {
    if (!connect_to_server()) {
        return;
    }

    AllocMessage msg;
    memset(&msg, 0, sizeof(msg));
    msg.type = AllocMessage::ALLOC;
    msg.device_id = device_id;
    msg.size = size;
    msg.buffer_type = buffer_type;
    msg.process_id = getpid();
    msg.buffer_id = buffer_id;
    msg.timestamp = std::chrono::system_clock::now().time_since_epoch().count();

    // Serialize socket writes to prevent message interleaving from multiple threads
    std::lock_guard<std::mutex> lock(socket_mutex_);

    // CRITICAL: Use blocking send to ensure message is fully delivered
    // MSG_DONTWAIT was causing dropped messages when socket buffer filled up
    size_t total_sent = 0;
    while (total_sent < sizeof(msg)) {
        ssize_t sent = send(
            socket_fd_,
            reinterpret_cast<const char*>(&msg) + total_sent,
            sizeof(msg) - total_sent,
            0);  // Blocking send
        if (sent < 0) {
            if (errno == EINTR) {
                // Interrupted by signal, retry
                continue;
            }
            // Connection lost, mark as disconnected
            connected_ = false;
            return;
        }
        total_sent += sent;
    }
}

void AllocationClient::send_deallocation_message(int device_id, uint64_t buffer_id) {
    if (!connect_to_server()) {
        return;
    }

    AllocMessage msg;
    memset(&msg, 0, sizeof(msg));
    msg.type = AllocMessage::FREE;
    msg.device_id = device_id;
    msg.buffer_id = buffer_id;
    msg.process_id = getpid();

    // Serialize socket writes to prevent message interleaving from multiple threads
    std::lock_guard<std::mutex> lock(socket_mutex_);

    // CRITICAL: Use blocking send to ensure message is fully delivered
    // MSG_DONTWAIT was causing dropped messages when socket buffer filled up
    size_t total_sent = 0;
    while (total_sent < sizeof(msg)) {
        ssize_t sent = send(
            socket_fd_,
            reinterpret_cast<const char*>(&msg) + total_sent,
            sizeof(msg) - total_sent,
            0);  // Blocking send
        if (sent < 0) {
            if (errno == EINTR) {
                // Interrupted by signal, retry
                continue;
            }
            // Connection lost, mark as disconnected
            connected_ = false;
            return;
        }
        total_sent += sent;
    }
}

// Circular buffer allocation/deallocation
void AllocationClient::send_cb_allocation_message(int device_id, uint64_t size, uint64_t cb_id) {
    if (!connect_to_server()) {
        return;
    }

    AllocMessage msg;
    memset(&msg, 0, sizeof(msg));
    msg.type = AllocMessage::CB_ALLOC;
    msg.device_id = device_id;
    msg.size = size;
    msg.buffer_id = cb_id;
    msg.process_id = getpid();
    msg.timestamp = std::chrono::system_clock::now().time_since_epoch().count();

    std::lock_guard<std::mutex> lock(socket_mutex_);
    size_t total_sent = 0;
    while (total_sent < sizeof(msg)) {
        ssize_t sent = send(socket_fd_, reinterpret_cast<const char*>(&msg) + total_sent, sizeof(msg) - total_sent, 0);
        if (sent < 0) {
            if (errno == EINTR) {
                continue;
            }
            connected_ = false;
            return;
        }
        total_sent += sent;
    }
}

void AllocationClient::send_cb_deallocation_message(int device_id, uint64_t size, uint64_t cb_id) {
    if (!connect_to_server()) {
        return;
    }

    AllocMessage msg;
    memset(&msg, 0, sizeof(msg));
    msg.type = AllocMessage::CB_FREE;
    msg.device_id = device_id;
    msg.size = size;  // Include size for proper deallocation tracking
    msg.buffer_id = cb_id;
    msg.process_id = getpid();

    std::lock_guard<std::mutex> lock(socket_mutex_);
    size_t total_sent = 0;
    while (total_sent < sizeof(msg)) {
        ssize_t sent = send(socket_fd_, reinterpret_cast<const char*>(&msg) + total_sent, sizeof(msg) - total_sent, 0);
        if (sent < 0) {
            if (errno == EINTR) {
                continue;
            }
            connected_ = false;
            return;
        }
        total_sent += sent;
    }
}

// Kernel load/unload
void AllocationClient::send_kernel_load_message(int device_id, uint64_t size, uint64_t kernel_id, uint8_t kernel_type) {
    if (!connect_to_server()) {
        return;
    }

    AllocMessage msg;
    memset(&msg, 0, sizeof(msg));
    msg.type = AllocMessage::KERNEL_LOAD;
    msg.device_id = device_id;
    msg.size = size;
    msg.buffer_id = kernel_id;
    msg.buffer_type = kernel_type;  // 0=App, 1=Fabric, 2=Dispatch
    msg.process_id = getpid();
    msg.timestamp = std::chrono::system_clock::now().time_since_epoch().count();

    std::lock_guard<std::mutex> lock(socket_mutex_);
    size_t total_sent = 0;
    while (total_sent < sizeof(msg)) {
        ssize_t sent = send(socket_fd_, reinterpret_cast<const char*>(&msg) + total_sent, sizeof(msg) - total_sent, 0);
        if (sent < 0) {
            if (errno == EINTR) {
                continue;
            }
            connected_ = false;
            return;
        }
        total_sent += sent;
    }
}

void AllocationClient::send_kernel_unload_message(int device_id, uint64_t size, uint64_t kernel_id) {
    if (!connect_to_server()) {
        return;
    }

    AllocMessage msg;
    memset(&msg, 0, sizeof(msg));
    msg.type = AllocMessage::KERNEL_UNLOAD;
    msg.device_id = device_id;
    msg.size = size;  // CRITICAL: Pass size for proper deallocation tracking
    msg.buffer_id = kernel_id;
    msg.process_id = getpid();

    std::lock_guard<std::mutex> lock(socket_mutex_);
    size_t total_sent = 0;
    while (total_sent < sizeof(msg)) {
        ssize_t sent = send(socket_fd_, reinterpret_cast<const char*>(&msg) + total_sent, sizeof(msg) - total_sent, 0);
        if (sent < 0) {
            if (errno == EINTR) {
                continue;
            }
            connected_ = false;
            return;
        }
        total_sent += sent;
    }
}

// Static interface methods
void AllocationClient::report_allocation(int device_id, uint64_t size, uint8_t buffer_type, uint64_t buffer_id) {
    auto& inst = instance();
    if (inst.enabled_) {
        inst.send_allocation_message(device_id, size, buffer_type, buffer_id);
    }
}

void AllocationClient::report_deallocation(int device_id, uint64_t buffer_id) {
    auto& inst = instance();
    if (inst.enabled_) {
        inst.send_deallocation_message(device_id, buffer_id);
    }
}

void AllocationClient::report_cb_allocation(int device_id, uint64_t size, uint64_t cb_id) {
    auto& inst = instance();
    if (inst.enabled_) {
        inst.send_cb_allocation_message(device_id, size, cb_id);
    }
}

void AllocationClient::report_cb_deallocation(int device_id, uint64_t size, uint64_t cb_id) {
    auto& inst = instance();
    if (inst.enabled_) {
        inst.send_cb_deallocation_message(device_id, size, cb_id);
    }
}

void AllocationClient::report_kernel_load(int device_id, uint64_t size, uint64_t kernel_id, uint8_t kernel_type) {
    auto& inst = instance();
    if (inst.enabled_) {
        inst.send_kernel_load_message(device_id, size, kernel_id, kernel_type);
    }
}

void AllocationClient::report_kernel_unload(int device_id, uint64_t size, uint64_t kernel_id) {
    auto& inst = instance();
    if (inst.enabled_) {
        inst.send_kernel_unload_message(device_id, size, kernel_id);
    }
}

bool AllocationClient::is_enabled() { return instance().enabled_; }

}  // namespace tt::tt_metal

File: tt_metal/programming_examples/alexp_examples/memory_utilization_monitor/allocation_client.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
# SPDX-License-Identifier: Apache-2.0

"""
Python client for Allocation Server

This demonstrates how Python applications can report allocations
to the allocation server for cross-process tracking.
"""

import socket
import struct
import os
import time
from enum import IntEnum


class MessageType(IntEnum):
    """Message types for allocation server protocol"""

    ALLOC = 1
    FREE = 2
    QUERY = 3
    RESPONSE = 4
    DUMP_REMAINING = 5
    DEVICE_INFO_QUERY = 6
    DEVICE_INFO_RESPONSE = 7


class BufferType(IntEnum):
    """Buffer types matching TT-Metal BufferType enum"""

    DRAM = 0
    L1 = 1
    L1_SMALL = 2
    TRACE = 3


class AllocationClient:
    """
    Client for reporting allocations to the allocation server.

    Usage:
        client = AllocationClient()
        buffer_id = client.allocate(device_id=0, size=1024*1024, buffer_type=BufferType.L1)
        # ... use buffer ...
        client.deallocate(buffer_id)
    """

    SOCKET_PATH = "/tmp/tt_allocation_server.sock"

    # Message format: Type(u8), device_id(i32), size(u64), buffer_type(u8),
    #                 process_id(i32), buffer_id(u64), timestamp(u64),
    #                 [response fields: 4x u64 + device info fields: 2x u64 + 6x u32]
    # Total: 1 + 3(pad) + 4 + 8 + 1 + 3(pad) + 4 + 8 + 8 + 32 + 16 + 24 = 112 bytes
    MESSAGE_FORMAT = "=BxxxiQBxxxiQQQQQQQQIIIIII"  # Little-endian

    def __init__(self):
        """Connect to allocation server"""
        self.sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        try:
            self.sock.connect(self.SOCKET_PATH)
            print(f"✓ Connected to allocation server")
        except Exception as e:
            raise RuntimeError(
                f"Failed to connect to allocation server: {e}\n" f"Make sure allocation_server_poc is running!"
            )

        self.next_buffer_id = 1
        self.process_id = os.getpid()

    def __del__(self):
        """Close connection"""
        if hasattr(self, "sock"):
            self.sock.close()

    def allocate(self, device_id: int, size: int, buffer_type: BufferType) -> int:
        """
        Report an allocation to the server.

        Args:
            device_id: Device ID (0-7)
            size: Size in bytes
            buffer_type: Type of buffer (DRAM, L1, etc.)

        Returns:
            buffer_id: Unique identifier for this allocation
        """
        buffer_id = self.next_buffer_id
        self.next_buffer_id += 1

        timestamp = int(time.time() * 1e9)  # nanoseconds

        msg = struct.pack(
            self.MESSAGE_FORMAT,
            MessageType.ALLOC,  # type
            device_id,  # device_id
            size,  # size
            buffer_type,  # buffer_type
            self.process_id,  # process_id
            buffer_id,  # buffer_id
            timestamp,  # timestamp
            0,
            0,
            0,
            0,  # response fields (unused)
            0,
            0,  # device info: total_dram_size, total_l1_size
            0,
            0,
            0,
            0,
            0,
            0,  # device info: arch_type, num_dram_channels, dram_size_per_channel, l1_size_per_core, is_available, num_devices
        )

        self.sock.send(msg)
        return buffer_id

    def deallocate(self, buffer_id: int):
        """
        Report a deallocation to the server.

        Args:
            buffer_id: The buffer ID returned from allocate()
        """
        msg = struct.pack(
            self.MESSAGE_FORMAT,
            MessageType.FREE,  # type
            0,  # device_id (unused)
            0,  # size (unused)
            0,  # buffer_type (unused)
            self.process_id,  # process_id
            buffer_id,  # buffer_id
            0,  # timestamp (unused)
            0,
            0,
            0,
            0,  # response fields (unused)
            0,
            0,  # device info fields (unused)
            0,
            0,
            0,
            0,
            0,
            0,  # device info fields (unused)
        )

        self.sock.send(msg)

    def query_device(self, device_id: int) -> dict:
        """
        Query memory statistics for a device.

        Args:
            device_id: Device ID to query

        Returns:
            dict with keys: dram_allocated, l1_allocated, l1_small_allocated,
                          trace_allocated, total_buffers
        """
        msg = struct.pack(
            self.MESSAGE_FORMAT,
            MessageType.QUERY,  # type
            device_id,  # device_id
            0,
            0,
            0,
            0,
            0,  # unused fields
            0,
            0,
            0,
            0,  # response fields (will be filled)
            0,
            0,  # device info fields (will be filled)
            0,
            0,
            0,
            0,
            0,
            0,  # device info fields (will be filled)
        )

        self.sock.send(msg)

        # Receive response
        response = self.sock.recv(struct.calcsize(self.MESSAGE_FORMAT))
        data = struct.unpack(self.MESSAGE_FORMAT, response)

        return {
            "device_id": data[1],
            "dram_allocated": data[7],
            "l1_allocated": data[8],
            "l1_small_allocated": data[9],
            "trace_allocated": data[10],
        }

    def get_num_devices(self) -> int:
        """
        Query the number of available devices.

        Returns:
            Number of devices detected by the server
        """
        msg = struct.pack(
            self.MESSAGE_FORMAT,
            MessageType.DEVICE_INFO_QUERY,  # type
            -1,  # device_id = -1 means query for device count
            0,
            0,
            0,
            0,
            0,  # unused fields
            0,
            0,
            0,
            0,  # response fields
            0,
            0,  # device info fields (will be filled)
            0,
            0,
            0,
            0,
            0,
            0,  # device info fields (will be filled)
        )

        self.sock.send(msg)

        # Receive response
        response = self.sock.recv(struct.calcsize(self.MESSAGE_FORMAT))
        data = struct.unpack(self.MESSAGE_FORMAT, response)

        # num_devices is at index 17 (last field)
        return data[17]

    def query_device_info(self, device_id: int) -> dict:
        """
        Query detailed device information.

        Args:
            device_id: Device ID to query (or -1 for device count)

        Returns:
            dict with device capabilities: arch_type, total_dram_size, total_l1_size,
            num_dram_channels, dram_size_per_channel, l1_size_per_core, is_available
        """
        msg = struct.pack(
            self.MESSAGE_FORMAT,
            MessageType.DEVICE_INFO_QUERY,  # type
            device_id,  # device_id
            0,
            0,
            0,
            0,
            0,  # unused fields
            0,
            0,
            0,
            0,  # response fields
            0,
            0,  # device info fields (will be filled)
            0,
            0,
            0,
            0,
            0,
            0,  # device info fields (will be filled)
        )

        self.sock.send(msg)

        # Receive response
        response = self.sock.recv(struct.calcsize(self.MESSAGE_FORMAT))
        data = struct.unpack(self.MESSAGE_FORMAT, response)

        arch_names = {0: "Invalid", 1: "Grayskull", 2: "Wormhole_B0", 3: "Blackhole", 4: "Quasar"}

        return {
            "device_id": data[1],
            "is_available": bool(data[16]),
            "arch_type": data[12],
            "arch_name": arch_names.get(data[12], "Unknown"),
            "total_dram_size": data[11],
            "total_l1_size": data[10],  # Note: indices shifted for Q vs I
            "num_dram_channels": data[13],
            "dram_size_per_channel": data[14],
            "l1_size_per_core": data[15],
            "num_devices": data[17],
        }


def format_bytes(bytes_val: int) -> str:
    """Format bytes in human-readable form"""
    units = ["B", "KB", "MB", "GB", "TB"]
    unit_idx = 0
    size = float(bytes_val)

    while size >= 1024.0 and unit_idx < len(units) - 1:
        size /= 1024.0
        unit_idx += 1

    return f"{size:.2f} {units[unit_idx]}"


def demo():
    """Demo showing how to use the allocation client"""
    print("\n🧪 Python Allocation Client Demo")
    print(f"   PID: {os.getpid()}")
    print(f"   Watch the server and monitor output!\n")

    try:
        client = AllocationClient()
        buffers = []

        # Step 1: Allocate some L1
        print("[Step 1] Allocating 4MB of L1...")
        for i in range(4):
            buffer_id = client.allocate(device_id=0, size=1024 * 1024, buffer_type=BufferType.L1)
            buffers.append(buffer_id)
            print(f"  Allocated buffer {buffer_id}")
        time.sleep(3)

        # Step 2: Allocate some DRAM
        print("\n[Step 2] Allocating 100MB of DRAM...")
        for i in range(4):
            buffer_id = client.allocate(device_id=0, size=25 * 1024 * 1024, buffer_type=BufferType.DRAM)
            buffers.append(buffer_id)
            print(f"  Allocated buffer {buffer_id}")
        time.sleep(3)

        # Step 3: Query stats
        print("\n[Step 3] Querying current statistics...")
        stats = client.query_device(device_id=0)
        print(f"  Device 0:")
        print(f"    DRAM: {format_bytes(stats['dram_allocated'])}")
        print(f"    L1: {format_bytes(stats['l1_allocated'])}")
        time.sleep(3)

        # Step 4: Free half
        print("\n[Step 4] Freeing half the buffers...")
        half = len(buffers) // 2
        for buffer_id in buffers[:half]:
            client.deallocate(buffer_id)
            print(f"  Freed buffer {buffer_id}")
        buffers = buffers[half:]
        time.sleep(3)

        # Step 5: Free all
        print("\n[Step 5] Freeing all remaining buffers...")
        for buffer_id in buffers:
            client.deallocate(buffer_id)
            print(f"  Freed buffer {buffer_id}")
        buffers.clear()
        time.sleep(2)

        print("\n✅ Demo complete!")

    except Exception as e:
        print(f"❌ Error: {e}")
        return 1

    return 0


if __name__ == "__main__":
    import sys

    sys.exit(demo())
